src/metabase/public_settings.clj

```diff
@@ -627,10 +630,23 @@
   :type       :boolean
   :default    false)
 
+(defn- not-handling-api-request?
+  []
+  (nil? @api/*current-user*))
+
+(defn set-uploads-database-id!
+  "Sets the :uploads-database-id setting, with an appropriate permission check."
+  [new-id]
+  (if (or (not-handling-api-request?)
+          (mi/can-write? Database new-id))
+    (setting/set-value-of-type! :integer :uploads-database-id new-id)
+    (api/throw-403)))
+
 (defsetting uploads-database-id
```

Comment:
I think this setting might fit better in an uploads namespace. `metabase.public-settings` is kind of a weird namespace; the name implies it should only be for _public_ settings but it's become a catch-all for random settings. I just raised this on Slack in #engineering-backend.

But settings in general aren't tied to their namespace in anyway so they can go wherever is most sensible.